tests/File/ReadCsvTest.php

```diff
@@ -0,0 +1,364 @@
+            ],
+            [
+                [
+                    '1,2,"3"',
+                    '2,"\n3",4',
+                    '"5",6',
+                ],
+                [
+                    ['1', '2', '3'],
+                    ['2', '\n3', '4'],
+                    ['5', '6'],
+                ],
+            ],
+            [
+                [
+                    '1,2,"3"',
+                    '2,"""3""",4',
+                    '"5",6',
+                ],
+                [
+                    ['1', '2', '3'],
+                    ['2', '"3"', '4'],
+                    ['5', '6'],
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForWithConfig
+     * @param        array $lines
+     * @param        array $config
+     * @param        array $expected
+     */
+    public function testWithConfig(array $lines, array $config, array $expected): void
+    {
+        // Given
+        $file = FileFixture::createFromLines($lines, $this->root->url());
+        $result = [];
+        [$separator, $enclosure, $escape] = $config;
+
+        // When
+        foreach (File::readCsv($file, $separator, $enclosure, $escape) as $datum) {
+            $result[] = $datum;
+        }
+
+        // Then
+        $this->assertEquals($expected, $result);
+    }
+
+    public function dataProviderForWithConfig(): array
+    {
+        return [
+            [
+                [],
+                [",", "\"", "\\"],
+                [],
+            ],
+            [
+                [],
+                [";", "\"", "\\"],
+                [],
+            ],
+            [
+                [],
+                [",", "'", "\\"],
+                [],
+            ],
+            [
+                [],
+                [",", "\"", "/"],
+                [],
+            ],
+            [
+                [],
+                [",", "\"", ""],
+                [],
+            ],
+            [
+                ['1,2,3'],
+                [",", "\"", "\\"],
+                [['1', '2', '3']],
+            ],
+            [
+                ['1;2;3'],
+                [",", "\"", "\\"],
+                [['1;2;3']],
+            ],
+            [
+                ['1,2,3'],
+                [";", "\"", "\\"],
+                [['1,2,3']],
+            ],
+            [
+                ['1;2;3'],
+                [";", "\"", "\\"],
+                [['1', '2', '3']],
+            ],
+            [
+                ["1,'2',3"],
+                [",", "'", "\\"],
+                [['1', '2', '3']],
+            ],
+            [
+                ["'1','2','3'"],
+                [",", "'", "\\"],
+                [['1', '2', '3']],
+            ],
+            [
+                [
+                    '1,2,3',
+                    '4,5,6',
+                ],
+                [",", "\"", "\\"],
+                [
+                    ['1', '2', '3'],
+                    ['4', '5', '6'],
```

Comment:
Suggestion: For unit tests, we want to try to avoid any unnecessary logic via code, as we want to test as straightforwardly as possible our code without introducing other avenues for mistakes/bugs to slip in via code our tests. Therefore, consider possibly using [Nowdoc](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.nowdoc) or Heredoc syntax to create multi-line strings for processing, rather than manipulating nested array structures to create a multiline file. Now/Heredoc is kind of verbose and often results in code that breaks indentation, but I think it is more important to have tests that have simple setups (Given sections) to ensure the test is easy to understand there is no question what it is doing.

Reply:
Hi @markrogoyski,

Do you think it's really necessary in this case? To be honest, I don't like this syntax precisely because it spoils the code design. It seems to me that there is quite a bit of logic in the construction of the fixture. But if you insist, I will of course rewrite using this syntax.

Reply:
Yeah, I agree it is ugly syntax, especially when it breaks indentation. No need to rewrite any existing tests.

Reply:
Thank you Mark!